helix-term/src/grammars.rs

```diff
@@ -0,0 +1,290 @@
+}
+
+fn run_parallel<F>(grammars: Vec<GrammarConfiguration>, job: F)
+where
+    F: Fn(GrammarConfiguration) + std::marker::Send + 'static + Copy,
+{
+    let mut n_jobs = 0;
+    let pool = threadpool::Builder::new().build();
+    let (tx, rx) = channel();
+
+    for grammar in grammars {
+        let tx = tx.clone();
+        n_jobs += 1;
+
+        pool.execute(move || {
+            job(grammar);
+
+            // report progress
+            tx.send(1).unwrap();
+        });
+    }
+    pool.join();
+
+    assert_eq!(rx.try_iter().sum::<usize>(), n_jobs);
+}
+
+pub fn fetch_grammar(grammar: GrammarConfiguration) {
+    if let GrammarSource::Git { remote, revision } = grammar.source {
+        let grammar_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
+            .join("../runtime/grammars/sources")
+            .join(grammar.grammar_id.clone());
+
+        fs::create_dir_all(grammar_dir.clone()).expect("Could not create grammar directory");
+
+        // create the grammar dir contains a git directory
+        if !grammar_dir.join(".git").is_dir() {
+            Command::new("git")
+                .args(["init"])
+                .current_dir(grammar_dir.clone())
+                .output()
+                .expect("Could not execute 'git'");
+        }
+
+        // ensure the remote matches the configured remote
+        if get_repository_info(&grammar_dir, vec!["remote", "get-url", REMOTE_NAME])
+            != Some(remote.clone())
+        {
+            set_remote(&grammar_dir, &remote);
+        }
+
+        // ensure the revision matches the configured revision
+        if get_repository_info(&grammar_dir, vec!["rev-parse", "HEAD"]) != Some(revision.clone()) {
+            // Fetch the exact revision from the remote.
+            // Supported by server-side git since v2.5.0 (July 2015),
+            // enabled by default on major git hosts.
+            Command::new("git")
+                .args(["fetch", REMOTE_NAME, &revision])
+                .current_dir(grammar_dir.clone())
+                .output()
+                .expect("Failed to execute 'git'");
+
+            Command::new("git")
+                .args(["checkout", &revision])
+                .current_dir(grammar_dir)
+                .output()
+                .expect("Failed to execute 'git'");
+
+            println!(
+                "Grammar '{}' checked out at '{}'.",
+                grammar.grammar_id, revision
+            );
+        } else {
+            println!("Grammar '{}' is already up to date.", grammar.grammar_id);
+        }
+    };
+}
+
+// Sets the remote for a repository to the given URL, creating the remote if
+// it does not yet exist.
+fn set_remote(repository: &Path, remote_url: &str) {
+    if !Command::new("git")
+        .args(["remote", "set-url", REMOTE_NAME, remote_url])
+        .current_dir(repository)
+        .output()
+        .expect("Failed to execute 'git'")
+        .status
+        .success()
```

Comment:
Is the only error condition for this command that the remote doesn't exist? Or are there others? Would it be better to do `git remote` and check if the remote name is in the output?

Reply:
I believe this can only fail if the remote doesn't exist yet. I'll check the docs :+1:

Reply:
It looks like there are [a few other error conditions](https://github.com/libgit2/libgit2/blob/1d811f0eb990ecbba406f9a7debf463669218e9d/src/remote.c#L628-L660) related to invalid names or git config. However, it doesn't look like there is a reasonable way to pick apart the error cases without pulling in libgit2, so I'm not sure if it's worth going that far. I guess we can bet on if there is some kind of config error that `git remote add` will fail for the same reason.